### PR TITLE
Mark source stages as finished when parent finishes

### PR DESCRIPTION
**Pull request by losipiuk**
In fault tolerant execution it was not enforced that source stages were marked as finished immediatelly after parent stages completes successfully. As a result the query which essentially compelted successfully could have been marked as failed if one of source stages failed.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix issue when queries could sometimes fail when run with `retry_policy` set to `TASK`. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java [hunk 1]
```diff
@@ -1952,6 +1958,8 @@ private StageExecution(
             this.taskSource = requireNonNull(taskSource, "taskSource is null");
             this.sinkPartitioningScheme = requireNonNull(sinkPartitioningScheme, "sinkPartitioningScheme is null");
             this.exchange = requireNonNull(exchange, "exchange is null");
+            requireNonNull(sourceStageExecutions, "sourceStageExecutions is null");
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java:1961]:
```suggestion
this.sourceStageExecutions = ImmutableMap.copyOf(requireNonNull(sourceStageExecutions, "sourceStageExecutions is null"));
```

clearly optional nit.

**Review by findinpath — APPROVED** [review]

**Comment by electrum** [comment]
Description has a couple of typos "immediatelly" and "compelted"

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java [hunk 2]
```diff
@@ -2380,6 +2388,22 @@ private void finish()
             stage.finish();
             taskSource.close();
 
+            sourceStageExecutions.forEach(StageExecution::markFinished);
+
+            recordFinishStats();
+        }
+
+        // parent already finished so no need to continue execution
+        private void markFinished()
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java:2397]: I see some duplication here with the above, but I'm not familiar with the structure of this to suggest a better approach. I'm assuming you considered this.

**Review by electrum — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java (hunk 2 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java:2397]: Yeah - I can parametrize. Or maybe name differently. The major difference is that standard `finish` assumes that for each partition processed by the stage there was a task which completed successfully. The `markFinished` is when there is still work to be done by tasks from that stage, but we no longer care.

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java]: Restructured a bit - hopefully more readable, and less duplication

**Pull request merged**